new/resources-anywhere.md

```diff
@@ -0,0 +1,211 @@
+- 99.99% of things that currently work in Chef 12, must continue to work in Chef 13.
+- As much as possible should work in 12.X, so that people in Chef 12 have a story when we move to 13 and so people can begin converting now.
+
+#### Accessing `Chef::Resource::File` or `Chef::Provider::File` directly
+
+In Chef 13, core resources are simply moving out of the Resource and Provider namespaces.  However, in Chef 12, there will be resources and providers who extend from them or construct them directly.
+
+In Chef 12, we will create classes in `Chef::Resource` extending from the real Resources which emit a deprecation warning on initialize or subclassing (`class Chef::Resource::File < Chef::Resources::File ...`).
+
+#### Accessing `File` from a resource or provider method
+
+Right now, when you type `File` in any resource or provider it always refers to `Chef::Resource::File` or `Chef::Provider::File`.  In Chef 13, it will refer to `::File`.
+
+The deprecation warnings for `Chef::Resource::File` suffice for Chef 12 deprecation purposes.
+
+Now for the reverse compatibility part: we want `File` to work.  We will allow anyone who declares a resource *outside* of Chef to type `File` normally by including Chef::Resource::TopLevelConstants directly into their resource.  Whenever a resource is created in `Chef::Resource`, we will check if there is a top level constant with the same name, and add it.  Since only resources *outside* of `Chef::Resource` get this included, anything inside `Chef::Resource` will still refer to the original thing.
+
+LWRPs created via `resources/x.rb` in a cookbook are also exempt from `TopLevelConstants`.
+
+#### Name Conflicts
+
+Resources can try to take the same name in recipe DSL (such as when two cookbooks both implement a similar thing).  Fact of life.  Presently when two custom resources try to take the same name, presently they will do one of two unsatisfactory things:
+
+- Raise a confusing "subclassing the wrong class" error on load; or
+- Silently overwrite and merge with one another, with the last one winning
+
+Now, when you are in separate namespaces, the first thing simply won't happen.
+
+We will avoid the second thing by checking whether the method `my_resource` is already defined, and if it is, we
+
+#### Replaced Classes
+
+It is possible to replace a class in Ruby (you can simply go assign ).  Due to the way this is implemented, the new class will in fact be used (which is sort of what a user would expect).  However, it's janky and something we'd like to warn about.  When the user attempts to use the resource (by typing `my_resource`), we will check in `build_resource` that the class has not changed from what was originally registered, and emit a warning that the class isn't what we expect and that they should contact the author of the new class to get the warning fixed.
+
+To *legitimately* do this, you should use `chef_recipe_dsl` instead.
```

Comment:
We already stopped doing this with Chef 12 and stopped undefining the constants and recreating classes.  This feels like we're reintroducing the same problem again.  Classes need to *not* be redefinable.  Allowing them to be redefined leads to tons of memory leaks in chefspec which runs mutiple converges rapidly in the same process.

I think this is also a solved problem with the resource resolver.   You can have multiple classes declare that they provide the same resource with `provides` lines, and then the user can determine via the resource_priority_map which class wins if there's a conflict.